Workaround for Boost compilation error with Clang 18 and libc++ in ossfuzz buildpack-deps

r0qs opened the pull request:
This PR implements the same workaround applied by the ossfuzz maintainers to the Solidity ossfuzz setup:

https://github.com/google/oss-fuzz/blob/5a9322260c4c66e7eb3e5d922b9fee6621ffd8da/projects/solidity/Dockerfile#L17-L18

It replaces `base-clang` with `base-builder`, the same image used by ossfuzz, that utilizes Clang 15. This change allows us to continue running our instance in CI until a fix for the main issue is available.

Workaround for: https://github.com/ethereum/solidity/issues/15711

r0qs committed:
Replace base-clang by ossfuzz base-builder

r0qs commented on scripts/docker/buildpack-deps/Dockerfile.ubuntu.clang.ossfuzz:
```diff
@@ -59,26 +62,19 @@ RUN apt-get update; \
 		pylint \
 		requests \
-		tabulate \
-		z3-solver;
```
Z3 is already installed from source in the steps below, so this is unnecessary. I also removed CMake, as it is already included in the base image.

blishko replied:
This is the python package of `Z3`, which is unrelated to the Z3 binary built below.
However, if it can be removed without anything breaking, that it's good to remove it :)

r0qs replied:
Yeah, it is not used by any CI job based on the ossfuzz image as far as I know. The [chk_proofs](https://github.com/ethereum/solidity/blob/71988f7546063a9d6d63d54ece85839400b46264/.circleci/config.yml#L943) which is the job that runs the script mentioned here: https://github.com/ethereum/solidity/pull/15551#discussion_r1821295419 uses the Ubuntu 24.04 image and it is not run by ossfuzz. This is why I'm also wondering if all the smt stuff is really needed here, or if we should just remove it from this image.

r0qs replied:
Ok, I updated Z3 and CVC5, but I kept the removal of the `z3-python`, I don't think it is relevant for this image. Because it is required by the tests under `test/formal`, which indeed uses the z3 python bindings, but I believe the fuzzer does not run such tests. Maybe there are more things that we could cleanup as well.

r0qs commented on scripts/docker/buildpack-deps/Dockerfile.ubuntu.clang.ossfuzz:
```diff
@@ -21,8 +21,10 @@
-LABEL version="8"
+# FIXME: Workaround for Boost 1.83.0 build failure when using Clang 18 (default since https://github.com/google/oss-fuzz/pull/11714)
+# See: https://github.com/google/oss-fuzz/blob/5a9322260c4c66e7eb3e5d922b9fee6621ffd8da/projects/solidity/Dockerfile#L17C1-L18C139
+FROM gcr.io/oss-fuzz-base/base-builder@sha256:19782f7fe8092843368894dbc471ce9b30dd6a2813946071a36e8b05f5b1e27e AS base
```
Hey @bshastry do you see any problem in switch images here?

nikola-matic reviewed (dismissed):
Soft approve. Let's wait until end of week for @bshastry to chime in regarding the base image change.

r0qs commented:
> Soft approve. Let's wait until end of week for @bshastry to chime in regarding the base image change.

Sure, better to have a response from him on that. Also, honestly, we could maybe even cleanup a bit more our current image. Do we fuzz with smt enabled?

r0qs committed:
Bump Z3 and CVC5

bshastry commented:
LGTM if build passes :-)

blishko commented:
> Ok, I updated Z3 and CVC5, but I kept the removal of the z3-python, I don't think it is relevant for this image. Because it is required by the tests under test/formal, which indeed uses the z3 python bindings, but I believe the fuzzer does not run such tests. Maybe there are more things that we could cleanup as well.

@r0qs, it would be better to unify it with the way I updated the other Docker images.
The libraries are not needed anymore, just executables. I also think you do not need to build Z3 from scratch, no? Just fetching the released executable should be OK, I think?

r0qs committed:
Convert spaces to tabs

r0qs commented:
> > Ok, I updated Z3 and CVC5, but I kept the removal of the z3-python, I don't think it is relevant for this image. Because it is required by the tests under test/formal, which indeed uses the z3 python bindings, but I believe the fuzzer does not run such tests. Maybe there are more things that we could cleanup as well.
> 
> @r0qs, it would be better to unify it with the way I updated the other Docker images. The libraries are not needed anymore, just executables. I also think you do not need to build Z3 from scratch, no? Just fetching the released executable should be OK, I think?

Yes, I guess just fetching the released binaries should be sufficient indeed. But notice that Z3 was build as static library, not sure if this was a requirement.

blishko reviewed (approved):
LGTM!
